arbor/backends/gpu/reduce_by_key.hpp

```diff
@@ -96,74 +46,73 @@ float get_from_lane(float value, unsigned lane) {
+    unsigned key_mask;
 
     __device__ __inline__
     bool is_root() const {
         return left == lane_id;
     }
 
-    __device__ __inline__
-    bool may_cross_warp() const {
-        return left==0 || right==threads_per_warp();
-    }
-
-    template <typename I1>
     __device__
-    run_length_type(I1 idx) {
-        auto right_limit = [] (unsigned roots, unsigned shift) {
+    run_length(int idx, unsigned mask) {
+        key_mask = mask;
+        lane_id = threadIdx.x%impl::threads_per_warp();
+        unsigned num_lanes = impl::threads_per_warp()-__clz(key_mask);
+
+        auto right_limit = [num_lanes] (unsigned roots, unsigned shift) {
             unsigned zeros_right  = __ffs(roots>>shift);
-            return zeros_right ? shift -1 + zeros_right : threads_per_warp();
+            return zeros_right ? shift -1 + zeros_right : num_lanes;
         };
 
-        lane_id = threadIdx.x%threads_per_warp();
-
-        // determine if this thread is the root
-        int left_idx  = arb_shfl_up(idx, 1);
+        // determine if this thread is the root (i.e. first thread with this key)
+        int left_idx  = __shfl_up_sync(key_mask, idx, lane_id? 1: 0);
         int is_root = 1;
         if(lane_id>0) {
             is_root = (left_idx != idx);
         }
 
         // determine the range this thread contributes to
-        unsigned roots = arb_ballot(is_root);
+        unsigned roots = __ballot_sync(key_mask, is_root);
 
+        // determine the bounds of the lanes with the same key as idx
         right = right_limit(roots, lane_id+1);
-        left  = threads_per_warp()-1-right_limit(__brev(roots), threads_per_warp()-1-lane_id);
-        width = rounddown_power_of_2(right - left);
+        left  = impl::threads_per_warp()-1-right_limit(__brev(roots), impl::threads_per_warp()-1-lane_id);
+
+        // find the largest power of two that is less than or equal to the run length
+        shift = impl::rounddown_power_of_2(right - left);
     }
 };
 
-} // namespace impl
-
 template <typename T, typename I>
 __device__ __inline__
-void reduce_by_key(T contribution, T* target, I idx) {
-    impl::run_length_type run(idx);
+void reduce_by_key(T contribution, T* target, I i, unsigned mask) {
+    run_length run(i, mask);
+    unsigned shift = run.shift;
+    const unsigned key_lane = run.lane_id - run.left;
 
-    // get local copies of right and width, which are modified in the reduction loop
-    auto rhs = run.right;
-    auto width = run.width;
+    bool participate = run.shift && run.lane_id+shift<run.right;
 
-    while (width) {
-        unsigned source_lane = run.lane_id + width;
+    while (__any_sync(run.key_mask, shift)) {
+        const unsigned w = participate? shift: 0;
+        const unsigned source_lane = run.lane_id + w;
 
-        auto source_value = impl::get_from_lane(contribution, source_lane);
-        if (source_lane<rhs) {
+        T source_value = __shfl_sync(run.key_mask, contribution, source_lane);
```

Comment:
It probably makes little difference, but I believe you could just use `__shfl_down_sync(..., w)` rather than compute `source_lane` explicitly.

Reply:
I think you are right.